### DISCUSSION TITLE
Documentation "Custom Request and APIRoute class" GZipMiddleware reference

### DISCUSSION THREAD

**Discussion by Tayum** [fastapi/fastapi#9104] [Questions]
The `Tip` in [Custom Request and APIRoute Class documentation | Create a custom GzipRequest class](https://fastapi.tiangolo.com/advanced/custom-request-and-route/#create-a-custom-gziprequest-class) says that the example shown there is a "toy" one, and for the real GZip support the [GZip Middleware](https://fastapi.tiangolo.com/advanced/middleware/#gzipmiddleware) could be used.

* The custom created GZip Request class handles the decoding of incoming gzipped content we **receive to a web-server**.
* But GZip Middleware is doing the opposite thing: it handles the GZip encoding of the content we **send from a web-server**.

1. Is the [suggested piece of information](https://fastapi.tiangolo.com/advanced/middleware/#gzipmiddleware) in "Tip" really corresponds to the [example](https://fastapi.tiangolo.com/advanced/custom-request-and-route/#create-a-custom-gziprequest-class) shown below in the documenation?   
2. Is there a plan to implement a GZip Middleware for decoding incoming gzipped requests (similarly to the existing GZip Middleware that encodes the outgoing responses)?

**Comment by jordan-cote** [comment]
It would indeed be great.

**Comment by jamescw19** [comment]
I came across this today and it's definitely confusing. The "Tip" gives the impression that `GZipMiddleware` handles both request and response compression.

Looking at similar issues for FastAPI/Starlette, it looks like there are no plans to implement request decompression in `GZipMiddleware` according to [this closed issue](https://github.com/encode/starlette/issues/644#issuecomment-549084132). It references a related [discussion on general-purpose deserialization](https://github.com/tiangolo/fastapi/issues/521#issuecomment-548950422) which contains an implementation of `GZipMiddleware` request decompression, but this hasn't been merged back into Starlette.

**Comment by harrv** [comment]
In my very first experience using FastAPI I ran into this issue. Incoming `POST` requests that have `Content-Encoding: gzip` fail request validation. Following the FastAPI doc to add the GZipMiddleware does not help my app to process incoming gzipped requests. From my testing, it appears that all the GZipMiddleware does with a request is to check if the `Accept-Encoding` header contains `gzip` and if so it gzips the response.

The [Starlette Middleware doc](https://www.starlette.io/middleware/#gzipmiddleware) also implies that GZipMiddleware is only for gzipping responses, not for handling gzipped requests.

Since this issue is still open, I thought I'd comment with this "me too" comment and suggest that unless there are plans to add request handling to the GZipMiddleware, then at a minimum the documentation should make clear that middleware will create gzipped responses but does not handle gzipped requests. Ideally that updated documentation would include an example of how to write a custom middleware to handle gzipped requests, if that is indeed the only way it can be done right now. The example could be either in-line in the doc, or linked from the doc.

Thanks!

Edit: To clarify, the documentation page linked to from the OP does already have an example of how to write a custom middleware that will handle gzipped POST requests. But people won't be looking for that unless the documentation page for GZipMiddleware makes it clear that it only gzips responses and you have to write a custom middleware if you want to handle gzipped POST requests. Then the page referenced in this issue could be linked to from the GZipMiddleware doc page.

**Comment by Kludex** [comment]
Yeah... The docs are wrong there... :thinking: 

I don't find anything on the literature about `Content-Encoding` in the request headers... :eyes:
